OMCompiler/SimulationRuntime/c/util/doubleEndedList.c

```diff
@@ -0,0 +1,534 @@
+ */
+void pushBackDoubleEndedList(DOUBLE_ENDED_LIST* list, const void* data) {
+  /* Error checking */
+  assertStreamPrint(NULL, 0 != list, "pushBackDoubleEndedList: invalid list-pointer");
+
+  /* Create new node */
+  DOUBLE_ENDED_LIST_NODE* newLastNode = createNodeDoubleEndedList(data, list->itemSize);
+
+  /* Append node at back */
+  if (list->length==0) {
+    list->first = newLastNode;
+    list->last = newLastNode;
+  } else {
+    list->last->next = newLastNode;
+    newLastNode->prev = list->last;
+    list->last = newLastNode;
+  }
+
+  list->length+= 1;
+}
+
+
+/**
+ * @brief Insert list element after given node.
+ *
+ * @param list        Pointer to double ended list.
+ * @param prevNode    Previous node for new created node.
+ * @param data        Pointer to data to be coppied into new node.
+ */
+void insertDoubleEndedList(DOUBLE_ENDED_LIST *list, DOUBLE_ENDED_LIST_NODE* prevNode, const void *data) {
+  /* Error checking */
+  assertStreamPrint(NULL, 0 != list, "insertDoubleEndedList: invalid list-pointer");
+  assertStreamPrint(NULL, 0 != prevNode, "insertDoubleEndedList: invalid previous-node-pointer");
+
+  /* Create new node */
+  DOUBLE_ENDED_LIST_NODE* newNode = createNodeDoubleEndedList(data, list->itemSize);
+
+  newNode->prev = prevNode;
+  newNode->next = prevNode->next;
+  prevNode->next = newNode;
+
+  /* Update end of list */
+  if(list->last == prevNode)
+    list->last = newNode;
+
+  list->length+= 1;
+}
+
+
+// ############################################################################
+//
+// Section for removing nodes
+//
+// ############################################################################
+
+
+/**
+ * @brief Remove single node from list.
+ *
+ * @param list    Double ended list.
+ * @param node    Node to be deleted from list.
+ */
+void removeNodeDoubleEndedList(DOUBLE_ENDED_LIST* list, DOUBLE_ENDED_LIST_NODE *node) {
+  if (node != NULL) {
+    /* Update previous node */
+    if (node->prev) {
+      if (node->next) {
+        node->prev->next = node->next;  /* Set next of previous node to be the node after deleted one */
+      } else {
+        node->prev->next = NULL;
+        if (node->next == NULL) { /* Previous node is now last node */
+          list->last = node->prev;
+        }
+      }
+    }
+    /* Update next node */
+    if (node->next) {
+      if (node->prev) {
+        node->next->prev = node->prev;  /* Set previous of next node to be the node before deleted one */
+      } else {
+        node->next->prev = NULL;
+        if (node->prev == NULL) { /* Next node is now first node */
+          list->first = node->next;
+        }
+      }
+    }
+
+    /* Free node */
+    freeNodeDoubleEndedList(node);
+    list->length -= 1;
+    if (list->length == 0) {
+      list->first = NULL;
+      list->last = NULL;
+    }
+  }
+}
+
+
+/**
+ * @brief Removes first node from list.
+ *
+ * @param list    Double ended list.
+ */
+void popFirstDoubleEndedList (DOUBLE_ENDED_LIST *list) {
```

Comment:
Maybe call this `removeFirstDoubleEndedList`. I think `pop` implies also returning the element, so this is a bit unexpected.

Reply:
Sounds like a good idea. I'll change it.